From dbb99d055289acffb72a07596e086ff114c76e2a Mon Sep 17 00:00:00 2001
From: Christian Marangi <ansuelsmth@gmail.com>
Date: Fri, 6 Dec 2024 16:09:44 +0100
Subject: [PATCH] iwinfo: limit assoclist client parsing element to
 IWINFO_BUFSIZE

Limit assoclist element to IWINFO_BUFSIZE. This is to fix a segfault if
the clients are more than 140. Once the local buffer is filled, the
function stop parsing additional clients and returns info for the parsed
ones.

Fixes: #15
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
---
 iwinfo_madwifi.c | 4 ++++
 iwinfo_nl80211.c | 7 +++++++
 iwinfo_nl80211.h | 1 +
 3 files changed, 12 insertions(+)

diff --git a/iwinfo_madwifi.c b/iwinfo_madwifi.c
index d27e6c9..a52d289 100644
--- a/iwinfo_madwifi.c
+++ b/iwinfo_madwifi.c
@@ -772,6 +772,10 @@ static int madwifi_get_assoclist(const char *ifname, char *buf, int *len)
 		do {
 			si = (struct ieee80211req_sta_info *) cp;
 
+			/* stop parsing more elements as we reached max buf */
+			if (bl + sizeof(entry) > IWINFO_BUFSIZE)
+				break;
+
 			memset(&entry, 0, sizeof(entry));
 
 			entry.signal = (si->isi_rssi - 95);
diff --git a/iwinfo_nl80211.c b/iwinfo_nl80211.c
index dc4ff93..16f5d31 100644
--- a/iwinfo_nl80211.c
+++ b/iwinfo_nl80211.c
@@ -2254,6 +2254,10 @@ static int nl80211_get_assoclist_cb(struct nl_msg *msg, void *arg)
 		[NL80211_RATE_INFO_SHORT_GI]     = { .type = NLA_FLAG   },
 	};
 
+	/* stop parsing more elements as we reached max buf */
+	if (arr->count >= arr->max)
+		return NL_STOP;
+
 	/* advance to end of array */
 	e += arr->count;
 	memset(e, 0, sizeof(*e));
@@ -2397,6 +2401,9 @@ static int nl80211_get_assoclist(const char *ifname, char *buf, int *len)
 	struct nl80211_array_buf arr = { .buf = buf, .count = 0 };
 	struct iwinfo_assoclist_entry *e;
 
+	/* Limit element to the preallocated space */
+	arr.max = IWINFO_BUFSIZE / sizeof(*e);
+
 	if ((d = opendir("/sys/class/net")) != NULL)
 	{
 		while ((de = readdir(d)) != NULL)
diff --git a/iwinfo_nl80211.h b/iwinfo_nl80211.h
index 2dff08b..fbbf2f7 100644
--- a/iwinfo_nl80211.h
+++ b/iwinfo_nl80211.h
@@ -67,6 +67,7 @@ struct nl80211_rssi_rate {
 struct nl80211_array_buf {
 	void *buf;
 	int count;
+	int max;
 };
 
 #endif
